CLUSTER.md

```diff
@@ -0,0 +1,451 @@
+consistent & timely manner.
+
+Maintaining these two tiers allows scaling the number of nodes while:
+- Maintaining an upper bound over the time required to decide and propagate failover decisions.
+While the time required to propagate the updated primary/replica status depends on the number of Failover Coordinators in the deployment
+the maximum number of these should always be small enough to allow us to guarantee a substantially better SLA than what
+the current implementation provides.
+- Limiting the clustering overhead on individual nodes in the deployment.
+Depending on pub/sub mode, an individual node only needs to maintain a connection to its Failover Coordinator (and primary/replicas).
+
+While Flotilla can be implemented using any sort of strongly consistent system, the reference implementation we propose
+uses Redis to host the Topology Director & Failover Coordinator logic. This allows for different deployment modes depending on 
+the needs and orchestrating abilities of the administrator:
+- Full Blown: Dedicated nodes holding the Topology Directors and multiple Failover Coordinators dividing the data nodes among them.
+- Midrange: Dedicated nodes holding both the Topology Director and Failover Coordinator (with just one Failover Coordinator owning all the data nodes).
+- Lightweight: Topology Director and Failover Coordinator located on a subset of data nodes. This removes the need to maintain and orchestrate
+non-data nodes at the cost of potentially impacting performance on the dual-use nodes.
+- Naive: Topology Director and Failover Coordinator located on all nodes. This allows for the quickest bootstrapping, but severely limits scalability.
+
+![Flotilla High Level overview](images/high_level.jpg?raw=true "High Level Overview")
+
+### The Topology Director
+
+The Topology Director serves as the source of truth for the cluster topology (excluding primary/replica status) and helps propagate
+information from Failover Coordinators. It never handles any user data, only topology metadata.
+Since it owns the metadata of the entire cluster it can allow changing this metadata while preserving the strong
+consistency guarantees we desire.
+For example - it can ensure that if a Shard holds a slot marked as Migrating, there's a corresponding Shard with the slot marked as Importing.
+We also rely on the Failover Coordinator to distribute the primary/replica state of individual nodes - This state is received periodically
+from every Failover Coordinator, cached in the Topology Director and relayed to all other Failover Coordinators when they send their state.
+
+- A completely reactive entity
+- Receives requests to alter the cluster topology and adjusts the internal state as needed
+- Validates change requests for correctness
+- Responds to topology queries by the other components
+
+#### Interfaces
+
+The Topology Director interface is fully declarative other than join/remove flows. This allows for easy orchestration via various tools.
```

Comment:
All of them are blocking.

Reply:
This will be a change from the cluster V1 where pretty much all cluster primitives are asynchronous. From an admin's perspective, going from async to sync is relatively straightforward but the reverse is not true. This would imply that the two protocols (V1 and V2) are not truly interchangeable, even if they have the exact same set of primitives, semantically speaking.

Reply:
They are definitely not interchangeable -- we explicitly specify (in the Summary section) we do not maintain compatibility with control path APIs like CLUSTER MEET / FORGET / SETSLOTS /....
The data path cluster interface (SLOTS / NODES / MOVED /... ) remains the same.